### ISSUE TITLE
Fully Convolutional Networks

### ISSUE THREAD

**Issue by adamjstewart** [qubvel/segmentation_models.pytorch#527]
Hi, I'm a developer for [TorchGeo](https://github.com/microsoft/torchgeo). In TorchGeo, we have a model for a [Fully Convolutional Network (FCN)](https://arxiv.org/abs/1411.4038). However, we're trying to keep TorchGeo specific to geospatial datasets/models/transforms, so we're interested in moving our FCN to this library. Would you be interested in this contribution? FCN's are a bit old for many computer vision tasks, but they often outperform U-Nets/Mask R-CNNs when working with satellite imagery since the entire receptive field isn't necessarily important.

Our current implementation looks like [this](https://github.com/microsoft/torchgeo/blob/main/torchgeo/models/fcn.py). If we were to contribute, would we just move this into a `segmentation_models_pytorch/fcn` directory? Are there any other files we would need to touch?

@calebrob6

**Comment by qubvel** [comment]
Hi @adamstewart 

Thanks for the proposal. I have some doubts regarding this architecture.
1. The architecture is very old and not widely used right now. Do you have any suggestions why anyone should use it? Does it have any advantages? Maybe just for research purposes as a baseline.
2. If FCN would be moved to SMP it should support different pretrained backbones (as all other architectures) and I am not sure this is what you want, because your implementation uses a custom and not pretrained backbone (which is not in SMP).

**Comment by adamjstewart** [comment]
1. FCN is state-of-the-art on many tasks in remote sensing. For example, see Figure 5 from https://arxiv.org/pdf/2101.01154.pdf where it is shown that FCN outperforms U-Net at resolving high-resolution features when trained on low-resolution labels.
2. I'm not sure how FCN would be modified to support different pretrained backbones since it was designed before backbones were common practice. If you can give me pointers I can try to implement it and make sure that it still gives us good performance.

**Comment by qubvel** [comment]
Hi @adamjstewart 

I might be wrong but your implementation is not actually FCN architecture, you have just 5 conv+leaky_relu layers without any downsampling and upsampling operation, while original architectures from the [paper](https://arxiv.org/abs/1411.4038) have 5 downsampling stages and several upsampling ones. Correct me if I'm wrong, please.

![Снимок экрана от 2021-12-30 18-14-29](https://user-images.githubusercontent.com/31920396/147764157-eb54a677-eb9f-4153-9924-d7543b14ac13.png)

**Comment by adamjstewart** [comment]
@calebrob6 wrote our FCN implementation, he might be better to answer.

**Comment by calebrob6** [comment]
@qubvel you are correct that our implementation is not from the FCN paper -- it is just _a_ "fully convolutional network" implementation (and considering this, I don't think I named our model well). The original point of this architecture was to have a small receptive field and we found it does interesting things when you are working on land cover mapping with weak supervision.

A way to achieve this in SMP might be something like a "simple" architecture that just consists of a stacked encoder/decoder as well as a convolution only encoder. Thoughts?
